### DISCUSSION TITLE
When are entity identifiers reused?

### DISCUSSION THREAD

**Discussion by JuanDiegoMontoya** [skypjack/entt#1205] [Q&A] [answered]
See title. I'm trying to evaluate how dangerous it is to have a component that references an entity with a different lifetime. I'd like to know when calling registry::valid on a destroyed entity will no longer return false. A simple test reveals that entities aren't reused *immediately*, but I don't know how reliable this behavior is:
```cpp
auto registry = entt::registry();
auto foo = registry.create();
registry_.destroy(foo);
for (int i = 0; i < 1'000'000; i++) {
    registry.create(); // Create a lot of entities.
}
assert(!registry.valid(foo)); // Okay, foo is still invalid.
```

In [this discussion](https://github.com/skypjack/entt/discussions/875) a similar problem is acknowledged, but it's still not clear to me what pattern will result in the reuse of an entity+version, nor if there's supposed to be a guarantee.

**Comment by skypjack** [comment, marked as answer]
Identifiers have an entity part and a version part. The version is incremented when you destroy it. The entity remains stable instead.
This means that the identifiers appears again in your game when the version wraps to 0 again. The default version for a 32b identifier is 12b. As anything else in EnTT, the version part as well as the type used for identifiers can be customized if you like.

**Reply by JuanDiegoMontoya** [reply]
Alright, thanks. It might be worth clarifying in the docs the mechanism by which entities are reused (e.g. they are added to a conceptual queue or stack) as it can impact when an identifier is reused. The best I could find on the topic was a mention of the version getting incremented upon release [in this section](https://github.com/skypjack/entt/wiki/Entity-Component-System#the-registry-the-entity-and-the-component).

As for my code, I'll probably add additional checks for the existence of components to avoid UB in the unlikely event that the stored identifier was destroyed and reused.

P.S. I changed my test slightly and I can see that, with the default settings, it reuses the identifier after the 4094th version when simply creating and destroying entities. For anyone else looking, [here's a link](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,selection:(endColumn:5,endLineNumber:14,positionColumn:5,positionLineNumber:14,selectionStartColumn:5,selectionStartLineNumber:14,startColumn:5,startLineNumber:14),source:'%23include+%3Centt/entt.hpp%3E%0A%0Aint+main()%0A%7B%0A++++auto+registry+%3D+entt::registry()%3B%0A++++auto+foo+%3D+registry.create()%3B%0A++++registry.destroy(foo)%3B%0A++++for+(int+i+%3D+0%3B+i+%3C+4097%3B+i%2B%2B)+%7B%0A++++++++auto+e+%3D+registry.create()%3B+//+Create+a+lot+of+entities.%0A++++++++printf(%22%25u,+%25u%5Cn%22,+entt::to_entity(e),+entt::to_version(e))%3B%0A++++++++if+(entt::to_version(e)+!!%3D+0)%0A++++++++++++registry.destroy(e)%3B%0A++++%7D%0A++++%0A++++printf(%22registry.valid(foo)+%3D+%25d%22,+(int)registry.valid(foo))%3B%0A%7D'),l:'5',n:'0',o:'C%2B%2B+source+%231',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((g:!((h:compiler,i:(compiler:g142,filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'0',intel:'0',libraryCode:'0',trim:'1',verboseDemangling:'0'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,libs:!((name:entt,ver:trunk)),options:'',overrides:!(),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+x86-64+gcc+14.2+(Editor+%231)',t:'0')),k:50,l:'4',m:50,n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compilerName:'x86-64+gcc+14.2',editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+x86-64+gcc+14.2+(Compiler+%231)',t:'0')),header:(),l:'4',m:50,n:'0',o:'',s:0,t:'0')),k:50,l:'3',n:'0',o:'',t:'0')),l:'2',n:'0',o:'',t:'0')),version:4) to the test on Compiler Explorer.
